### PR TITLE
feat: inject dapp templates

### PR DESCRIPTION
**Pull request by PainterPuppets**
feat: inject dapp templates

 - [x] create a dapp sample project
 - [x] implement init command

image
image

### CODE REVIEW

File: packages/core/src/project-creation.ts [hunk 1]
```diff
@@ -0,0 +1,148 @@
...
+    .filter((f) => fsExtra.pathExistsSync(f))
+    .map((f) => path.relative(process.cwd(), f));
+
+  if (existingFiles.length > 0) {
+    const errorMsg = `We couldn't initialize the sample project because these files already exist: ${existingFiles.join(
+      ', ',
+    )}
+
+Please delete or move them and try again.`;
+    console.log(console.log(errorMsg));
```
Comment (yanguoyu) [inline packages/core/src/project-creation.ts:76]:
```suggestion
    console.error(console.log(errorMsg));
```
Or we can define consistent rules for the logs later.

File: packages/core/src/project-creation.ts [hunk 2]
```diff
@@ -0,0 +1,148 @@
...
+    {
+      name: 'kuai-project',
+    },
+    { spaces: 2 },
+  );
+}
+
+export async function createProject(): Promise<void> {
+  if (!(await fsExtra.pathExists('package.json'))) {
+    await createPackageJson();
```
Comment (yanguoyu) [inline packages/core/src/project-creation.ts:111]: Maybe we can copy the `package.json` from sample-projects, and replace the version and name.

**Review by yanguoyu — COMMENTED** [review]
How will users use this to create a simple project?

**Comment by yanguoyu** [comment]
> > How will users use this to create a simple project?
> 
> just typing `kuai init`

Most of the time, if I use `npx kuai init`, I will not create a package.json first. So I think the package.json also needs to copy from `sample-projects`

File: packages/core/package.json [hunk 3]
```diff
@@ -8,6 +8,10 @@
     "test": "jest"
   },
   "dependencies": {
+    "enquirer": "2.3.6",
     "find-up": "2.1.0"
+  },
+  "devDependencies": {
+    "@types/fs-extra": "^9.0.13"
```
Comment (Keith-CY) [inline packages/core/package.json:15]: Dependency should be pinned

File: packages/core/recommended-gitignore.txt [hunk 4]
```diff
@@ -0,0 +1,8 @@
+node_modules
+.env
```
Comment (Keith-CY) [inline packages/core/recommended-gitignore.txt:2]: `.env.example` should be excluded for demonstration, or submit the `.env` and ignore `.env.local`, `.env.production` etc.

Comment (Keith-CY) [reply packages/core/recommended-gitignore.txt:2]: `.gitignore` generated by [gitignore](https://www.toptal.com/developers/gitignore/) would help, e.g. https://www.toptal.com/developers/gitignore/api/dotenv,node,windows,macos,linux,git

File: packages/core/sample-projects/tsconfig.json [hunk 5]
```diff
@@ -0,0 +1,10 @@
+{
+  "compilerOptions": {
+    "target": "es2020",
```
Comment (Keith-CY) [inline packages/core/sample-projects/tsconfig.json:3]: Is is recommended to target `es2020` by default?

File: packages/core/src/project-creation.ts [hunk 6]
```diff
@@ -0,0 +1,148 @@
...
+    const errorMsg = `We couldn't initialize the sample project because these files already exist: ${existingFiles.join(
+      ', ',
+    )}
+
+Please delete or move them and try again.`;
+    console.log(console.log(errorMsg));
+    process.exit(1);
+  }
+
+  await fsExtra.copy(path.join(packageRoot, 'sample-projects'), projectRoot);
```
Comment (Keith-CY) [inline packages/core/src/project-creation.ts:80]: Copying the directory could be done by [fs.cpSync(src, dest[, options])](https://nodejs.org/api/fs.html#fscpsyncsrc-dest-options) with `option.recursive: true` and `fs-extra` could be removed

File: packages/core/src/project-structure.ts [hunk 7]
```diff
@@ -17,3 +20,10 @@ export function getUserConfigPath(): string | undefined {
 
   return pathToConfigFile;
 }
+
+export async function getRecommendedGitIgnore(): Promise<string> {
+  const packageRoot = getPackageRoot();
+  const gitIgnorePath = path.join(packageRoot, 'recommended-gitignore.txt');
+
+  return fsExtra.readFile(gitIgnorePath, 'utf-8');
```
Comment (Keith-CY) [inline packages/core/src/project-structure.ts:28]: Does `fs.readFileSync` work here?

File: packages/core/src/util/packageInfo.ts [hunk 8]
```diff
@@ -0,0 +1,22 @@
...
+
+export function findClosestPackageJson(file: string): string | undefined {
+  return findup.sync('package.json', { cwd: path.dirname(file) });
+}
+
+export function getPackageJsonPath(): string {
+  const packageJsonPath = findClosestPackageJson(__filename);
+
+  if (!packageJsonPath) {
+    throw new Error('not found package.json');
```
Comment (Keith-CY) [inline packages/core/src/util/packageInfo.ts:12]: The message could be a bit detailed and `not found package.json` should be `package.json not found`

File: packages/core/src/project-creation.ts [hunk 9]
```diff
@@ -0,0 +1,145 @@
...
+
+async function addGitIgnore(projectRoot: string) {
+  const gitIgnorePath = path.join(projectRoot, '.gitignore')
+
+  let content = await getRecommendedGitIgnore()
+
+  if (fs.existsSync(gitIgnorePath)) {
+    const existingContent = fs.readFileSync(gitIgnorePath, 'utf-8')
+    content = `${existingContent}
+${content}`
```
Comment (Keith-CY) [inline packages/core/src/project-creation.ts:92]: `fs.appendFileSync` should work here

Ref: [fs.appendFileSync](https://nodejs.org/api/fs.html#fsappendfilesyncpath-data-options)

File: packages/core/src/project-creation.ts [hunk 10]
```diff
@@ -0,0 +1,145 @@
...
+  const { projectRoot, shouldAddGitIgnore } = responses
+
+  if (shouldAddGitIgnore) {
+    await addGitIgnore(projectRoot)
+  }
+
+  await copySampleProject(projectRoot)
+
+  console.log(`✨ Project created ✨\n`)
+  console.log('See the README.md file for some example tasks you can run\n')
```
Comment (Keith-CY) [inline packages/core/src/project-creation.ts:144]: Use `console.info` instead of `console.log` for prompting messages

**Comment by PainterPuppets** [comment]
Optimize code via conversations
add a `printAsciiLogo` & `printWelcomeMessage` function, now look like this
image

File: packages/core/src/project-creation.ts [hunk 11]
```diff
@@ -0,0 +1,125 @@
...
+}
+
+async function addGitIgnore(projectRoot: string) {
+  const gitIgnorePath = path.join(projectRoot, '.gitignore')
+
+  const content = await getRecommendedGitIgnore()
+
+  if (fs.existsSync(gitIgnorePath)) {
+    fs.appendFileSync(gitIgnorePath, content)
+  }
```
Comment (Keith-CY) [inline packages/core/src/project-creation.ts:61]: `return` is required in this branch, or `fs.writeFileSync` will be called after `fs.appendFileSync`

**Comment by Keith-CY** [comment]
Any update on this PR?

**Review by yanguoyu — APPROVED** [review]

**Review by Keith-CY — APPROVED** [review]

**Comment by Keith-CY** [comment]
This PR has been approved by 2 reviewers and can be merged after conflicts are resolved.

**Pull request merged**